Feature/bjs2 66747

Artpyak opened the pull request:
Выполнение задачи https://faang-school.atlassian.net/browse/BJS2-65874
и выполнение задачи https://faang-school.atlassian.net/browse/BJS2-66747

Artpyak committed:
BJS2-65874
1. Реализована отправка запроса на менторство.
2. Реализованно получение всех запросов на менторство с фильтрацией.
3. Реализовано принятие запроса на менторство.
4. Реализованна возможность отклонить запрос на менторство.

Artpyak committed:
BJS2-65874
1. Правки yaml

Artpyak committed:
BJS2-65874
1. Правки в MentorshipRequestRepository.java
Добавил RETURNING *

Artpyak committed:
BJS2-66747
Написание юнит-тестов

Artpyak committed:
BJS2-66747
Написание юнит-тестов

Artpyak committed:
Update MentorshipRequestRepository.java

Artpyak committed:
Update MentorshipRequestRepository.java

safon4ik10 commented on src/main/java/school/faang/user_service/controller/MentorshipRequestController.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.List;
+
+@Slf4j
```
Зачем тут логгер если логов нет?

Artpyak replied:
Убрал.

safon4ik10 commented on src/main/java/school/faang/user_service/service/MentorshipRequestServiceImpl.java:
```diff
@@ -0,0 +1,144 @@
+                                    .getCreatedAt()
+                                    .plusMonths(minRequestIntervalInMonths)
+                                    .isAfter(LocalDateTime.now())) {
+                                validateRequestInterval();
```
Не думаю что это корректное название метода, он ничего не валидирует

Artpyak replied:
Изменил на более корректное.

Artpyak committed:
BJS2-66747
1. Удалён неиспользуемый метод в маппере.
2. Изменено название метода на более корректное.
3. Удалена неиспользуемая аннотация.

Artpyak committed:
Merge remote-tracking branch 'origin/feature/BJS2-66747' into feature/BJS2-66747

# Conflicts:
#	src/main/java/school/faang/user_service/controller/MentorshipRequestController.java
#	src/main/java/school/faang/user_service/service/MentorshipRequestServiceImpl.java

Graccio commented on src/test/java/school/faang/user_service/controller/MentorshipRequestControllerTest.java:
```diff
@@ -0,0 +1,83 @@
+    void testNullDescription() {
+        MentorshipRequestDto mentorshipRequestDto = new MentorshipRequestDto();
+
+        assertThrows(
```
Ожидать здесь RuntimeException кажется не совсем правильным, так как у него есть несколько наследников, а в оригинальном методе выбрасывается NullPointerException с сообщением - предлагаю уточнить ожидаемый класс исключения + проверить, что полученное сообщение будет соответствовать ожидаемому с помощью assertEquals(expectedMessage, exception.getMessage())

Artpyak replied:
Исправил

Graccio commented on src/test/java/school/faang/user_service/controller/MentorshipRequestControllerTest.java:
```diff
@@ -0,0 +1,83 @@
+
+        mentorshipRequestController.requestMentorship(mentorshipRequestDto);
+
+        Mockito.verify(mentorshipRequestService, Mockito.times(1))
```
Как предложение - использовать статический импорт для методов Mockito. Так и код будет проще писать, и читать его станет удобнее

Artpyak replied:
Сделал. Действительно удобнее.

Graccio commented on src/test/java/school/faang/user_service/service/TestMentorshipMentorshipRequestDescriptionFilter.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.stream.Stream;
+
+public class TestMentorshipMentorshipRequestDescriptionFilter implements MentorshipRequestFilter {
```
Так и задумывалось назвать класс? И если это тест на фильтр, то может лучше переложить его в package filter?

Artpyak replied:
Убрал повторяющееся слово Mentorship из названия. 
Это тестовый фильтр который нужен для тестирования методов testGetRequests в сервисе. По этому я их и оставил в папке с сервисами. Но лучше сделаю для них отдельную папку в папке сервис.

Artpyak committed:
BJS2-66747
1. Переписал testNullDescription
2. Добавил статический импорт для методов Mockito.
3. Тестовые фильтры поместил в отдельную папку в папке сервиса.
4. Поработал над названиями тестовых методов.

PavelEzhkov commented on src/main/java/school/faang/user_service/dto/MentorshipRequestDto.java:
```diff
@@ -0,0 +1,18 @@
+    private String description;
+    private Long requesterId;
+    private Long receiverId;
+    private RequestStatus status;
```
Я предлагаю и для статусов делать дто

Artpyak replied:
Сделал.

PavelEzhkov commented on src/main/java/school/faang/user_service/dto/RejectionDto.java:
```diff
@@ -0,0 +1,6 @@
+
+import jakarta.validation.constraints.NotBlank;
+
+public record RejectionDto(@NotBlank String reason) {
```
А почему тут рекорд а везде ломбок и дата?

Artpyak replied:
Теперь ломбок и дата

PavelEzhkov commented on src/main/java/school/faang/user_service/filter/MentorshipRequestDescriptionFilter.java:
```diff
@@ -0,0 +1,24 @@
+package school.faang.user_service.filter;
```
Давай вынесем все папку mentorship или как тебе нравится, а то уже достаточно много фильтров в одной папке

Artpyak replied:
Вынес.

PavelEzhkov commented on src/main/java/school/faang/user_service/filter/MentorshipRequestFilter.java:
```diff
@@ -0,0 +1,13 @@
+public interface MentorshipRequestFilter {
+    boolean isApplicable(RequestFilterDto requestFilterDto);
+
+    Stream<MentorshipRequest> apply(
```
предлагаю обсудить переносы на стендап, но такой перенос не очень нравится мне, кажется что длинны строки хватает

PavelEzhkov commented on src/main/java/school/faang/user_service/service/MentorshipRequestServiceImpl.java:
```diff
@@ -0,0 +1,144 @@
+    private final UserRepository userRepository;
+    private final List<MentorshipRequestFilter> mentorshipRequestFilters;
+
+    @Value("${app.mentorship-request.min-request-interval-in-months}")
```
классно придумал через проперти файл

PavelEzhkov commented on src/main/java/school/faang/user_service/service/MentorshipRequestServiceImpl.java:
```diff
@@ -0,0 +1,144 @@
+                                    .getCreatedAt()
+                                    .plusMonths(minRequestIntervalInMonths)
+                                    .isAfter(LocalDateTime.now())) {
+                                throwIfRequestIntervalTooShort();
```
я бы всю проверку вынес тогда в приватный метод или же кидал ошибку тут, не то чтобы это не правильно, но такие реализации я никогда не видел, возможно чисто стилистически непривычно

Artpyak replied:
Вынес

PavelEzhkov commented on src/main/java/school/faang/user_service/service/MentorshipRequestServiceImpl.java:
```diff
@@ -0,0 +1,144 @@
+                            }
+                        })
+                );
+        checkingForDifferentIDs(requesterId, receiverId);
```
кажется что это проверку надо делать до похода в бд в начале метода

Artpyak replied:
Сделал

PavelEzhkov commented on src/main/java/school/faang/user_service/service/MentorshipRequestServiceImpl.java:
```diff
@@ -0,0 +1,144 @@
+
+    @Override
+    public List<MentorshipRequestDto> getRequests(RequestFilterDto requestFilterDto) {
+        Iterable<MentorshipRequest> allRequests = mentorshipRequestRepository.findAll();
+        Stream<MentorshipRequest> filteredMentorshipRequests =
+                ((Collection<MentorshipRequest>) allRequests).stream();
```
```suggestion
        Stream<MentorshipRequest> filteredMentorshipRequests =
                StreamSupport.stream(mentorshipRequestRepository.findAll().spliterator(), false);
```

PavelEzhkov commented on src/main/java/school/faang/user_service/service/MentorshipRequestServiceImpl.java:
```diff
@@ -0,0 +1,144 @@
+                            User requester = mentorshipRequest.getRequester();
+                            User receiver = mentorshipRequest.getReceiver();
+
+                            throwIfUserAlreadyInList(requester.getMentors(),
```
я бы исправил название метода throwIfUserAlreadyInList тк его цель не кидать ошибки, а проверять есть ли пользователи в списке. И еще мне не нравится, что передается 3 параметра, и часть из них только ради логов и текста ошибки

Artpyak replied:
Исправил

Artpyak committed:
BJS2-66747
1. Добавил RequestStatusMapper и внедрил его в MentorshipRequestMapper
2. Добавил в RequestStatusMapper метод requestStatusDtoToRequestStatus

Artpyak committed:
BJS2-66747
1. Исправил тесты в MentorshipRequestServiceImplTest
2. исправил RequestStatusMapper

Artpyak committed:
BJS2-66747
1. Исправил тесты MentorshipRequestServiceImplTest и MentorshipRequestStatusFilterTest
2. Исправил RequestFilterDto
3. Исправил MentorshipRequestStatusFilter

Artpyak committed:
BJS2-66747
1. Сделал RejectionDto обычным классом и исправил все баги связанные с этим.

Artpyak committed:
BJS2-66747
1. Вынес в приватный метод проверку интервала времени.
2. Обновил переносы строк.
3. Изменил метод по проверке на наличие в листе и переименовал его в validateUserNotInList.
